src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -90,99 +72,147 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol
+            }
+
+            return new TupleTypeSymbol(locations, tupleCompatibleType, elementLocations, elementNames, elementTypes);
+        }
+
+        private static NamedTypeSymbol EnsureRestExtensionsAreTuples(NamedTypeSymbol tupleCompatibleType)
+        {
+            if (!tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics[RestPosition - 1].IsTupleType)
             {
-                TupleFieldSymbol originalField = originalTuple._fields[fieldIndex];
-                TypeSymbol fieldType = currentLinkType.TypeArgumentsNoUseSiteDiagnostics[remainder];
-                fieldsBuilder.Add(originalField.WithType(this, currentLinkType, fieldType));
+                var nonTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
 
-                fieldIndex++;
-                if (fieldIndex >= originalTuple._fields.Length)
+                NamedTypeSymbol currentType = tupleCompatibleType;
+                do
                 {
-                    break;
+                    nonTupleTypeChain.Add(currentType);
+                    currentType = (NamedTypeSymbol)currentType.TypeArgumentsNoUseSiteDiagnostics[RestPosition - 1];
                 }
+                while (currentType.Arity == RestPosition);
 
-                remainder = fieldIndex % (RestPosition - 1);
-                if (remainder == 0)
+                if (!currentType.IsTupleType)
                 {
-                    currentLinkType = (NamedTypeSymbol)currentLinkType.TypeArgumentsNoUseSiteDiagnostics[RestPosition - 1];
+                    nonTupleTypeChain.Add(currentType);
                 }
+
+                Debug.Assert(nonTupleTypeChain.Count > 1);
+                tupleCompatibleType = nonTupleTypeChain.Pop();
+
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithModifiers>.GetInstance(RestPosition);
+
+                do
+                {
+                    var extensionTuple = Create(null, tupleCompatibleType, default(ImmutableArray<Location>), default(ImmutableArray<string>));
+                    tupleCompatibleType = nonTupleTypeChain.Pop();
+
+                    tupleCompatibleType = ReplaceRestExtensionType(tupleCompatibleType, typeArgumentsBuilder, extensionTuple);
+                }
+                while (nonTupleTypeChain.Count != 0);
+
+                typeArgumentsBuilder.Free();
+                nonTupleTypeChain.Free();
             }
 
-            _fields = fieldsBuilder.ToImmutableAndFree();
+            return tupleCompatibleType;
         }
 
-        /// <summary>
-        /// Copy the original tuple, but modify it to use new field names.
-        /// </summary>
-        private TupleTypeSymbol(TupleTypeSymbol originalTuple,
-                                ImmutableArray<string> newElementNames)
+        private static NamedTypeSymbol ReplaceRestExtensionType(NamedTypeSymbol tupleCompatibleType, ArrayBuilder<TypeWithModifiers> typeArgumentsBuilder, TupleTypeSymbol extensionTuple)
```

Comment:
Yes it does, there is an extra work + synchronization with other threads involved in getting new builder

---
In reply to: [61809481](https://github.com/dotnet/roslyn/pull/11006#discussion_r61809481) [](ancestors = 61809481)